File: backend/app/api/routes/poincare.py
"""
Poincaré hyperbolic embeddings endpoint for hierarchical skill knowledge graphs.

Maps the DAG ontology into the 2D Poincaré disk (||(u, v)|| < 1) where depth
translates directly to hyperbolic radius r = tanh(alpha * depth), preserving tree distances.
"""
from __future__ import annotations

import json
import math
from functools import lru_cache

from fastapi import APIRouter
from pydantic import BaseModel

from app.core.config import DATA_DIR

router = APIRouter(prefix="/poincare")


class PoincareNode(BaseModel):
    id: str
    name: str
    topic: str
    depth: int
    fan_out: int
    u: float
    v: float
    radius: float
    angle: float


class PoincareEdge(BaseModel):
    source: str
    target: str
    hyperbolic_dist: float


class PoincareResponse(BaseModel):
    nodes: list[PoincareNode]
    edges: list[PoincareEdge]


def _hyperbolic_dist(u1: float, v1: float, u2: float, v2: float) -> float:
    """Poincaré disk metric: d_H(x, y) = arcosh(1 + 2 * ||x-y||^2 / ((1-||x||^2)(1-||y||^2)))."""
    norm1_sq = u1 * u1 + v1 * v1
    norm2_sq = u2 * u2 + v2 * v2
    diff_sq = (u1 - u2) ** 2 + (v1 - v2) ** 2

    denom = (1.0 - norm1_sq) * (1.0 - norm2_sq)
    if denom <= 1e-9:
        return 5.0

    delta = 1.0 + 2.0 * diff_sq / denom
    delta = max(1.0, delta)
    return float(math.acosh(delta))


@router.get("", response_model=PoincareResponse)
async def get_poincare_disk_layout() -> PoincareResponse:
    """
    Compute 2D Poincaré disk embeddings for the entire skill graph ontology.
    """
    rows = json.loads((DATA_DIR / "skills.json").read_text(encoding="utf-8"))

    # Group skills by topic to assign angular sectors
    topics = sorted(list({r["topic"] for r in rows}))
    topic_angle_base = {t: (2 * math.pi * i) / len(topics) for i, t in enumerate(topics)}
    topic_counts: dict[str, int] = {}

    nodes: list[PoincareNode] = []
    node_map: dict[str, PoincareNode] = {}

    for r in rows:
        topic = r["topic"]
        depth = r["depth"]
        fan_out = r["fan_out"]

        # Hyperbolic radius r in [0, 0.90] based on depth
        r_hyperbolic = math.tanh(0.32 * depth)
        r_hyperbolic = min(0.88, r_hyperbolic)

        # Angular distribution within topic sector
        idx_in_topic = topic_counts.get(topic, 0)
        topic_counts[topic] = idx_in_topic + 1
        sector_span = (2 * math.pi / len(topics)) * 0.8
        angle = topic_angle_base[topic] + (idx_in_topic * 0.23) % sector_span

        u = round(r_hyperbolic * math.cos(angle), 4)
        v = round(r_hyperbolic * math.sin(angle), 4)

        node = PoincareNode(
            id=r["id"],
            name=r["name"],
            topic=topic,
            depth=depth,
            fan_out=fan_out,
            u=u,
            v=v,
            radius=round(r_hyperbolic, 3),
            angle=round(angle, 3),
        )
        nodes.append(node)
        node_map[node.id] = node

    # Compute edges with hyperbolic distances
    edges: list[PoincareEdge] = []
    for r in rows:
        target_id = r["id"]
        target_node = node_map.get(target_id)
        if not target_node:
            continue
        for prereq_id in r.get("prerequisites", []):
            source_node = node_map.get(prereq_id)
            if source_node:
                d_h = _hyperbolic_dist(
                    source_node.u, source_node.v, target_node.u, target_node.v
                )
                edges.append(
                    PoincareEdge(
                        source=prereq_id,
                        target=target_id,
                        hyperbolic_dist=round(d_h, 3),
                    )
                )

    return PoincareResponse(nodes=nodes, edges=edges)

File: lib/poincare.ts
import type { GraphNode, GraphEdge } from "./skillGraph";

export interface PoincareNode {
  id: string;
  name: string;
  category: string;
  depth: number;
  fanOut: number;
  u: number;
  v: number;
  radius: number;
  angle: number;
}

export interface PoincareEdge {
  from: string;
  to: string;
  hyperbolicDist: number;
}

export interface PoincareGraph {
  nodes: PoincareNode[];
  edges: PoincareEdge[];
}

/**
 * Poincaré disk metric:
 * d_H(x, y) = arcosh(1 + 2 * ||x-y||^2 / ((1-||x||^2)(1-||y||^2)))
 */
export function hyperbolicDist(u1: number, v1: number, u2: number, v2: number): number {
  const norm1Sq = u1 * u1 + v1 * v1;
  const norm2Sq = u2 * u2 + v2 * v2;
  const diffSq = (u1 - u2) ** 2 + (v1 - v2) ** 2;

  const denom = (1.0 - norm1Sq) * (1.0 - norm2Sq);
  if (denom <= 1e-9) {
    return 5.0;
  }

  const delta = Math.max(1.0, 1.0 + (2.0 * diffSq) / denom);
  return Math.acosh(delta);
}

/**
 * Compute 2D Poincaré disk embeddings for a skill graph DAG without node collisions.
 * Distributes nodes evenly along concentric depth orbits with angular separation.
 */
export function computePoincareLayout(
  nodes: GraphNode[],
  edges: GraphEdge[]
): PoincareGraph {
  // 1. Group nodes by depth levels
  const depthGroups = new Map<number, GraphNode[]>();
  for (const n of nodes) {
    const group = depthGroups.get(n.depth) || [];
    group.push(n);
    depthGroups.set(n.depth, group);
  }

  // 2. Count fanOut for each node
  const fanOutMap = new Map<string, number>();
  for (const edge of edges) {
    fanOutMap.set(edge.from, (fanOutMap.get(edge.from) ?? 0) + 1);
  }

  // 3. Define radius for each depth level (concentric horizons)
  // depth 0: 0.22, depth 1: 0.46, depth 2: 0.68, depth 3+: 0.86
  const depthRadii = [0.24, 0.48, 0.70, 0.86];

  const poincareNodes: PoincareNode[] = [];
  const nodeMap = new Map<string, PoincareNode>();

  // Sort depth levels
  const sortedDepths = Array.from(depthGroups.keys()).sort((a, b) => a - b);

  for (const depth of sortedDepths) {
    const group = depthGroups.get(depth) || [];
    const count = group.length;
    const r = depthRadii[Math.min(depth, depthRadii.length - 1)];

    // Stagger angle phase for each depth level to avoid vertical alignments
    const phaseOffset = depth * 0.45;

    group.forEach((n, idx) => {
      // Calculate evenly spaced angle across 360 degrees
      const angle = (2 * Math.PI * idx) / Math.max(1, count) + phaseOffset;

      const u = Number((r * Math.cos(angle)).toFixed(4));
      const v = Number((r * Math.sin(angle)).toFixed(4));

      const pNode: PoincareNode = {
        id: n.id,
        name: n.name,
        category: n.category,
        depth,
        fanOut: fanOutMap.get(n.id) ?? 0,
        u,
        v,
        radius: Number(r.toFixed(3)),
        angle: Number(angle.toFixed(3)),
      };

      poincareNodes.push(pNode);
      nodeMap.set(pNode.id, pNode);
    });
  }

  // 4. Build edge distances
  const poincareEdges: PoincareEdge[] = [];
  for (const e of edges) {
    const src = nodeMap.get(e.from);
    const tgt = nodeMap.get(e.to);
    if (!src || !tgt) continue;

    const dist = hyperbolicDist(src.u, src.v, tgt.u, tgt.v);
    poincareEdges.push({
      from: e.from,
      to: e.to,
      hyperbolicDist: Number(dist.toFixed(3)),
    });
  }

  return {
    nodes: poincareNodes,
    edges: poincareEdges,
  };
}
